src/NHibernate.Test/Linq/ParameterTypeLocatorTests.cs

```diff
@@ -0,0 +1,427 @@
+				{
+					{"2", o => o is EnumStoredAsStringType},
+					{"Large", o => o is EnumStoredAsStringType}
+				},
+				db.Users.Where(o => (o.NullableEnum1 ?? EnumStoredAsString.Large) == EnumStoredAsString.Medium),
+				db.Users.Where(o => EnumStoredAsString.Medium == (o.NullableEnum1 ?? EnumStoredAsString.Large))
+			);
+		}
+
+		[Test]
+		public void DoubleCoalesceTest()
+		{
+			AssertResults(
+				new Dictionary<string, Predicate<IType>>
+				{
+					{"2", o => o is EnumStoredAsStringType},
+					{"Large", o => o is EnumStoredAsStringType},
+				},
+				db.Users.Where(o => ((o.NullableEnum1 ?? (EnumStoredAsString?) EnumStoredAsString.Large) ?? o.Enum1) == EnumStoredAsString.Medium),
+				db.Users.Where(o => EnumStoredAsString.Medium == ((o.NullableEnum1 ?? (EnumStoredAsString?) EnumStoredAsString.Large) ?? o.Enum1))
+			);
+		}
+
+		[Test]
+		public void ConditionalTest()
+		{
+			AssertResults(
+				new Dictionary<string, Predicate<IType>>
+				{
+					{"2", o => o is EnumStoredAsStringType},
+					{"Unspecified", o => o is EnumStoredAsStringType}
+				},
+				db.Users.Where(o => (o.NullableEnum2.HasValue ? o.Enum1 : EnumStoredAsString.Unspecified) == EnumStoredAsString.Medium),
+				db.Users.Where(o => EnumStoredAsString.Medium == (o.NullableEnum2.HasValue ? EnumStoredAsString.Unspecified : o.Enum1))
+			);
+		}
+
+		[Test]
+		public void DoubleConditionalTest()
+		{
+			AssertResults(
+				new Dictionary<string, Predicate<IType>>
+				{
+					{"0", o => o is PersistentEnumType},
+					{"2", o => o is EnumStoredAsStringType},
+					{"Small", o => o is EnumStoredAsStringType},
+					{"Unspecified", o => o is EnumStoredAsStringType}
+				},
+				db.Users.Where(o => (o.Enum2 != EnumStoredAsInt32.Unspecified
+										? (o.NullableEnum2.HasValue ? o.Enum1 : EnumStoredAsString.Unspecified)
+										: EnumStoredAsString.Small) == EnumStoredAsString.Medium),
+				db.Users.Where(o => EnumStoredAsString.Medium == (o.Enum2 != EnumStoredAsInt32.Unspecified
+										? EnumStoredAsString.Small
+										: (o.NullableEnum2.HasValue ? EnumStoredAsString.Unspecified : o.Enum1)))
+			);
+		}
+
+		[Test]
+		public void CoalesceMemberTest()
+		{
+			AssertResults(
+				new Dictionary<string, Predicate<IType>>
+				{
+					{"2", o => o is EnumStoredAsStringType}
+				},
+				db.Users.Where(o => (o.NotMappedUser ?? o).Enum1 == EnumStoredAsString.Medium),
+				db.Users.Where(o => EnumStoredAsString.Medium == (o ?? o.NotMappedUser).Enum1)
+			);
+		}
+
+		[Test]
+		public void ConditionalMemberTest()
+		{
+			AssertResults(
+				new Dictionary<string, Predicate<IType>>
+				{
+					{"2", o => o is EnumStoredAsStringType},
+					{"\"test\"", o => o is AnsiStringType},
+				},
+				db.Users.Where(o => (o.Name == "test" ? o.NotMappedUser : o).Enum1 == EnumStoredAsString.Medium),
+				db.Users.Where(o => EnumStoredAsString.Medium == (o.Name == "test" ? o : o.NotMappedUser).Enum1)
+			);
+		}
+
+		[Test]
+		public void DynamicMemberTest()
```

Comment:
This test is failing.

```
Invalid type, actual type: p1
  Expected: True
  But was:  False

   at NHibernate.Test.Linq.ParameterTypeLocatorTests.AssertResult(Dictionary`2 expectedResults, QueryMode queryMode, Expression expression, Type targetType) in D:\BuildAgent\work\30546188361a242\src\NHibernate.Test\Linq\ParameterTypeLocatorTests.cs:line 418
   at NHibernate.Test.Linq.ParameterTypeLocatorTests.AssertResult(Dictionary`2 expectedResults, IQueryable query) in D:\BuildAgent\work\30546188361a242\src\NHibernate.Test\Linq\ParameterTypeLocatorTests.cs:line 369
   at NHibernate.Test.Linq.ParameterTypeLocatorTests.AssertResults(Dictionary`2 expectedResults, IQueryable[] queries) in D:\BuildAgent\work\30546188361a242\src\NHibernate.Test\Linq\ParameterTypeLocatorTests.cs:line 361
   at NHibernate.Test.Linq.ParameterTypeLocatorTests.DynamicMemberTest() in D:\BuildAgent\work\30546188361a242\src\NHibernate.Test\Linq\ParameterTypeLocatorTests.cs:line 233
```

And it seems the assert message does surface the expected information, yielding the parameter name instead of its type.

Reply:
> This test is failing.

I realized that visiting a dynamic expression in .NET Framework does not reduce it where on .NET Core it does, unless `DynamicExpressionVisitor` is used.

> And it seems the assert message does surface the expected information

Fixed.